### PR TITLE
Fix GH-29442 DataFrame.groupby doesn't preserve _metadata

### LINKED ISSUE #29442: DataFrame.groupby doesn't preserve _metadata
I'm following the 0.25.3 [documentation](https://pandas.pydata.org/pandas-docs/stable/development/extending.html#subclassing-pandas-data-structures) on using _metadata.

```python
sdf = SubclassedDataFrame2(my_data_as_dictionary)
sdf.added_property = 'hello pandas'
df = sdf.groupby('mycategorical')[['myfloat1', 'myfloat2']].sum()
print(df.added_property)
```
The above output produces the result:
AttributeError: 'DataFrame' object has no attribute 'added_property'

**added_property** is _not_ being "passed to manipulation results" as described in the documentation.

### PR DESCRIPTION
**Pull request by Japanuspus**
This bug is a regression in v1.1.0 and was introduced by the fix for GH-34214 in commit [6f065b].

Underlying cause is that the `*Splitter` classes do not use the `._constructor` property and do not call `__finalize__`.

Please note that the method name used for `__finalize__` calls was my best guess since documentation for the value has been hard to find.

[6f065b]: https://github.com/pandas-dev/pandas/commit/6f065b6d423ea211d803e8be93c27f547541c372

- [x] closes #29442
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by Japanuspus** [comment]
@jbrockmendel This PR is a two-line edit to fix a metadata regression in your recent work on groupby-performance. Would you be available to review? 

I am available to implement this is differently if needed.

File: pandas/tests/groupby/test_custom_metadata.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    @staticmethod
+    def from_table_data(df: pd.DataFrame, metadata) -> "PandasTable":
+        df = PandasTable(df)
+        object.__setattr__(df, _TABLE_METADATA_FIELD_NAME, metadata)
+        return df
+
+
+@pytest.fixture
+def dft():
+    df = pd.DataFrame([[11, 12, 0], [21, 22, 0], [31, 32, 1]], columns={"a", "b", "g"})
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_custom_metadata.py:74]: having columns be a set is really unusual, can you make this a list

File: pandas/tests/groupby/test_custom_metadata.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def _combine_metadata(data: List[str]) -> str:
+    """
+    A mock implementation for testing
+    """
+    return "+".join(data)
+
+
+class PandasTable(pd.DataFrame):
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_custom_metadata.py:28]: can this be tested without a subclass?  just set _metadata on a regular dataframe?

**Comment by jbrockmendel** [comment]
cc @TomAugspurger since this is about `__finalize__` being called.  I'm ambivalent, since this is Technically Correct, but also a place where we're really trying to avoid overhead since these methods get called in a loop.

**Comment by TomAugspurger** [comment]
At a glance, this feels like too low of a level to call `__finalize__`. Why can't we call it on the outer level, in whatever function is defining `DataFrameGroupBy.sum()`?

File: pandas/tests/groupby/test_custom_metadata.py (hunk 2 above)
Comment (Japanuspus) [reply pandas/tests/groupby/test_custom_metadata.py:28]: I don't think so: The `__finalize__` implementation in NDFrame copies all fields from `self._metadata`, but since self is a new object this would do nothing unless `_metadata` has been set for the class. We could make a test using a monkeypatched `DataFrame` class, but that would not catch the error of using `DataFrame` constructor rather than calling `_constructor`.

I could have a go at a more minimal test where only subclass features are `_metadata` and `_constructor` ?

**Comment by Japanuspus** [comment]
Thank you both for looking into this! I will start with a more concise testcase.

Should I also have a go at moving the `__finalize__` call to an outer layer, or are either of you looking into that in detail?

File: pandas/tests/groupby/test_custom_metadata.py (hunk 1 above)
Comment (Japanuspus) [reply pandas/tests/groupby/test_custom_metadata.py]: :blush: (converted from a dictionary init rather late in the process...). Is fixed in the updated concise testcase.

File: pandas/tests/groupby/test_custom_metadata.py (hunk 2 above)
Comment (Japanuspus) [reply pandas/tests/groupby/test_custom_metadata.py]: Have updated the testcase to use a minimal DataFrame extension. This still catches the original regression.

**Comment by Japanuspus** [comment]
Regarding performance implications of `__finalize__` in `_chop`

I tried running the benchmark from [GH-34214 ](https://github.com/pandas-dev/pandas/pull/34214), but this seems to hit `.fast_apply` and not `._chop`.

Can either of you come up with a benchmark that hits `._chop` in a hot loop where calling `__finalize__` on the groups would really be wasted?

(`fast_apply` benchmark for reference)
```
import numpy as np
from pandas import DataFrame

N = 10 ** 4
labels = np.random.randint(0, 2000, size=N)
labels2 = np.random.randint(0, 3, size=N)
df = DataFrame(
    {
        "key": labels,
        "key2": labels2,
        "value1": np.random.randn(N),
        "value2": ["foo", "bar", "baz", "qux"] * (N // 4),
    }
)

%timeit df.groupby(["key", "key2"]).apply(lambda x: 1)
%timeit df.groupby("key").apply(lambda x: 1)
```

**Comment by jbrockmendel** [comment]
> Can either of you come up with a benchmark that hits ._chop in a hot loop where calling __finalize__ on the groups would really be wasted?

Try setting `df.index = pd.CategoricalIndex(df.index)`

**Comment by Japanuspus** [comment]
Just running `df.index = pd.CategoricalIndex(df.index)` for the `df` produced in the `fast_apply` benchmark does not hit `_.chop`. But I have a feeling that was maybe not what you meant?

**Comment by TomAugspurger** [comment]
Perhaps run the groupby with a CategoricalIndex?

**Comment by Japanuspus** [comment]
Thanks for helping out! Managed to get a benchmark with some degradation:

With `df` as above
```
df.index = pd.CategoricalIndex(df.key)
%timeit df.groupby(level='key').apply(lambda x: 1)
```

Result was a 4% performance hit:
- This patch: 85.3 ms ± 570 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
- 1.1.x from source: 81.9 ms ± 1.48 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

I will not have time to learn enough of the inner workings of pandas to improve this on my own, but am willing to help out if you have some concrete pointers.  
Alternatively you could accept as-is, and then have a look at performance once tests are in place? Or just drop this (I can copy-paste the testcase to the issue thread) -- for my own use case I have implemented a fallback to handle initialization without `__finalize__`, so all is good for me.

File: pandas/tests/groupby/test_custom_metadata.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    @property
+    def _constructor(self):
+        return CustomDataFrame
+
+
+def test_groupby_with_custom_metadata():
+    custom_df = CustomDataFrame(
+        [[11, 12, 0], [21, 22, 0], [31, 32, 1]], columns=["a", "b", "g"]
+    )
+    custom_df._custom_metadata = "Custom metadata"
```
Comment (jreback) [inline pandas/tests/groupby/test_custom_metadata.py:22]: can you assert that the OP is working as well: https://github.com/pandas-dev/pandas/issues/29442

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_custom_metadata.py (hunk 3 above)
Comment (Japanuspus) [reply pandas/tests/groupby/test_custom_metadata.py]: Well spotted -- must admit I did not run the code from the original issue since it was missing inputs. After creating some input and adding the test, I realized that the issue I originally fixed was entirely unrelated...

In any case, I have also proposed a solution for the `.sum()` issue as well. Will resolve this comment, please open a new discussion directly on the proposed solution https://github.com/pandas-dev/pandas/pull/35688/commits/8d3d896bcb20f7bcb0d3ac2b71098dfb05fff11e.

If these two issues should be split into two pull requests, please let me know.

**Comment by TomAugspurger** [comment]
I guess one way to get a sense for the performance impact is to count how many times `__finalize__` is called. I think ideally it's called exactly once, rather than once per group.

**Comment by Japanuspus** [comment]
Good point @TomAugspurger.

With respect to my first commit, adding `__finalize__` in `DataSplitter._chop`, it seems this could be improved for the code paths tested here by moving `__finalize__` out to `_GroupBy.__iter__` and `_GroupBy.apply`. Should I make an attempt to do so, or would you rather take another approach? 

With respect to the `sum`-test, I could not see how to move the call  further out than `_agg_general` without having to duplicate it across all the aggregation methods? https://github.com/pandas-dev/pandas/pull/35688/commits/8d3d896bcb20f7bcb0d3ac2b71098dfb05fff11e

Also seems I have a bunch of failing tests, so will take a look at that. Please let me know if you would rather take this in another direction.

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
does this need a release note?

**Comment by Japanuspus** [comment]
@simonjayhawkins Yes, I should probably add a release note -- but there was a bit of scope creep (my original PR addressed my own problem, not the issue listed), and that work broke something. Also, I have come around to agreeing that we need a more performant version of my initial fix.

I will be short on time for a week or so, but then I will have a go at this.

**Comment by Japanuspus** [comment]
Just a note (mostly to self) on the failing tests since it took me an embarrassing amount of time to identify them.

The relevant tests are in `tests/generic/test_finalize.py` and are marked [`xfail`](https://docs.pytest.org/en/reorganize-docs/new-docs/user/xfail.html), which means that when running tests in strict mode, it will be reported as an error if they pass.
Since they include a test for aggregation by sum, they now pass which caused the test to be red in strict mode.

I will remove the `xfail` flag from the relevant cases, and then proceed to look at the performance as discussed above.

**Comment by jbrockmendel** [comment]
> The relevant tests are in tests/generic/test_finalize.py and are marked xfail

Tracking down xfailed tests that can be fixed is really helpful, thanks.

**Comment by Japanuspus** [comment]
This got sort of messy, and I think a summary is in place:

My initial PR did not address the linked issue and would cause performance regressions. When this was keenly pointed, I added a test and fix for the original issue.

After an unfortunate pause I continued to identify and disable x-failing tests caused by the fix, and also added a test for performance, counting the number of `__finalize__` calls in a `groupby().apply()` invocation, which shares relevant code-path with `.sum()`.

Finally, I have reverted the calls to `__finalize__` from my original PR since this is really an unrelated issue. I have left the test (with an `xfail`-tag), but will file a separate issue for this.

File: doc/source/whatsnew/v1.1.3.rst [hunk 4]
```diff
@@ -58,6 +58,7 @@ Bug fixes
 - Bug in :meth:`Series.astype` showing too much precision when casting from ``np.float32`` to string dtype (:issue:`36451`)
 - Bug in :meth:`Series.isin` and :meth:`DataFrame.isin` when using ``NaN`` and a row length above 1,000,000 (:issue:`22205`)
 - Bug in :func:`cut` raising a ``ValueError`` when passed a :class:`Series` of labels with ``ordered=False`` (:issue:`36603`)
+- Bug in :class:`BaseGroupedBy` causing ``groupby(...).sum()`` and similar to not preserve metadata (:issue:`29442`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.3.rst:61]: move to 1.1.4

File: pandas/tests/groupby/test_custom_metadata.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    # normal properties
+    _metadata = ["added_property"]
+
+    @property
+    def _constructor(self):
+        return SubclassedDataFrame2
+
+
+@pytest.mark.xfail(
+    reason="Missing high-performance implementation of metadata support for groupby. "
```
Comment (jreback) [inline pandas/tests/groupby/test_custom_metadata.py:23]: what is the issue here?

**Review by jreback — CHANGES_REQUESTED** [review]
this looks fine; i doubt performance is actually an issue here, can you show a benchmark otherwise?

File: pandas/tests/groupby/test_custom_metadata.py (hunk 5 above)
Comment (Japanuspus) [reply pandas/tests/groupby/test_custom_metadata.py:23]: TLDR: I plan to submit a separate PR to lift this x-fail which is unrelated to the linked issue.

The initial commit in this PR (7cc4d5375e57277f316c5a8bb17a58ab8003c5f1) included a call to `__finalize__` in the concrete `DataSplitter`-classes, but as pointed out by @TomAugspurger  and @jbrockmendel , this led to performance-degradation for aggregation code-paths due to unneeded calls to `__finalize__` for the intermediate dataframe/series objects. This is what is tested for in the performance test added in a2bdc3d3dfe80018bb6cabce3583fdf9df6cdca9.

Since calling `__finalize__` in `DataSplitter` turned out to be irrelevant for the linked issue (which was fixed by adding a `__finalize__` call in `BaseGroupBy._agg_general`, see 8d3d896bcb20f7bcb0d3ac2b71098dfb05fff11e) it has been removed from this PR.

One solution to lift the x-fail would be to call `__finalize__` for the groups returned by the iterator returned by `BaseGroup.get_iterator` (see 0af015ba77d8ff05f0d38a3d4b0518bbc737d393 for an implementation on my fork). I am getting some spurious test failures on this branch.

The reason I have not included this solution in the present PR is that I would rather submit it in a separate PR: Calling `__finalize__` independently from object initialization could potentially lead to duplicate calls to `__finalize__` for the same object, and I do not have sufficient overview of the code so see if this is the case in the proposed implementation. My thought was that a separate PR would make it easier to get an overview of such issues -- but I would of course be happy to merge into this PR if you prefer.

**Comment by Japanuspus** [comment]
> 
> this looks fine; i doubt performance is actually an issue here, can you show a benchmark otherwise?

For the PR as it stands, I don't think performance can be improved.

For the, unrelated and now removed, change in `DataSplitter`  (7cc4d5375e57277f316c5a8bb17a58ab8003c5f1) there is a benchmark in the comments above where we saw a 4% penalty. The alternative solution in 0af015ba77d8ff05f0d38a3d4b0518bbc737d393 only performs the `__finalize__` after aggregation and has no performance issues, but is of course less solid in the sense that there could be code-paths with multiple calls to `__finalize__`. This is the part I propose to move to another PR.

The applied benchmark for reference: 
```
import numpy as np
import pandas as pd

N = 10 ** 4
labels = np.random.randint(0, 2000, size=N)
labels2 = np.random.randint(0, 3, size=N)
df = pd.DataFrame(
    {
        "key": labels,
        "key2": labels2,
        "value1": np.random.randn(N),
        "value2": ["foo", "bar", "baz", "qux"] * (N // 4),
    }
)
df.index = pd.CategoricalIndex(df.key)
%timeit df.groupby(level='key').apply(lambda x: 1)

```

**Comment by Japanuspus** [comment]
> 
> 
> does this need a release note?

Added a release note for 1.1.4

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@Japanuspus pls limit changes to one thing in a PR, esp as this is a backport.

File: doc/source/whatsnew/v1.1.4.rst [hunk 6]
```diff
@@ -22,7 +22,7 @@ Fixed regressions
 
 Bug fixes
 ~~~~~~~~~
--
+- Bug in :class:`BaseGroupedBy` causing ``groupby(...).sum()`` and similar to not preserve metadata (:issue:`29442`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.4.rst:25]: you can remove the reference to BaseGroupedBy (this is not a real class) and these are user facing notes.

File: pandas/tests/groupby/test_custom_metadata.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            return super().__finalize__(*args, **kwargs)
+
+    N = 10 ** 4
+    labels = np.random.randint(0, N // 5, size=N)
+    labels2 = np.random.randint(0, 3, size=N)
+    df = SubclassedDataFrame3(
+        {
+            "key": labels,
+            "key2": labels2,
+            "value1": np.random.randn(N),
```
Comment (jreback) [inline pandas/tests/groupby/test_custom_metadata.py:59]: this file can be removed (as its not germane to the bug fix right)? if its a followon that's fine, but do in a separate PR targeting master.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_custom_metadata.py (hunk 7 above)
Comment (Japanuspus) [reply pandas/tests/groupby/test_custom_metadata.py]: The key test was a regression test for the linked issue, but there should be full coverage even without this test.

**Comment by Japanuspus** [comment]
I am slightly confused about @jreback mentioning this being a backport PR -- my intention was to target master. Or does this just mean that the patch will both apply to master and release branches?

**Comment by jreback** [comment]
> I am slightly confused about @jreback mentioning this being a backport PR -- my intention was to target master. Or does this just mean that the patch will both apply to master and release branches?

this si a regresion, so we target master, then its backported. so pls limit the change to the regression. if you want to do another change that targets master only that's ok, but separate PR

**Comment by Japanuspus** [comment]
> > I am slightly confused about @jreback mentioning this being a backport PR -- my intention was to target master. Or does this just mean that the patch will both apply to master and release branches?
> 
> this si a regresion, so we target master, then its backported. so pls limit the change to the regression. if you want to do another change that targets master only that's ok, but separate PR

Thank you for the explanation.
I believe the current PR is cut as tight as possible -- will go ahead with a separate PR for the `iter(...groupby())` issue.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @TomAugspurger if any comments.

**Comment by jreback** [comment]
thanks @Japanuspus
